nft_ingester/src/program_transformers/bubblegum/db.rs

```diff
@@ -114,42 +111,157 @@ where
     //TODO -> set maximum size of path and break into multiple statements
 }
 
-pub async fn update_asset<T>(
+pub async fn upsert_asset_with_leaf_info<T>(
```

Comment:
Why do we need to split the upserts between leaf / owner+delgate? Can't we handle it the same via a single upsert?

To attempt to answer my own question:

I assume the problem is that the seq value for a newer txn will be higher than the mint seq, meaning that some of the mint info will never get added to the asset table. So we need to break apart our seq value to distinguish between different attribute updates.

Reply:
Yes exactly right.  In order to handle out-of-order we need to break apart `seq` value for each "block" of data that can be updated by different instructions.

To use your example, the `seq` value for a `transfer` will be higher than the `seq` value when the asset was minted.  So if it was all upserted but controlled by the same `seq`, and if we saw the `transfer` come through first, then we would get the `leaf`, `owner`, and `delegate` from the `transfer` event, but we would never get the `specification_asset_class` which is something only set during `mint`.

That is why `mint` first does all the mint-only data, then `leaf`, `owner`, and `delegate` are separate.

Then to address why `leaf` is separate than `owner`+`delegate`,  it is because `redeem` and `decompress` update the `leaf` but do not update the `owner` and `delegate`.  So for example if `redeem` was indexed before a `transfer` that actually preceeded it on chain, and if they were all controlled by the same `seq`, then the `owner` and `delegate` would not be updated to their correct values.  `owner` and `delegate` can be grouped together because they are always updated as a pair.

This same phenomenon is true with a block of data I am calling "compression info" which is updated by both `mint` and `decompress`.  And this phenomenon is even true with the base `seq` number itself, if we want to keep that accurate in the asset table.

The general form of this whole PR's approach, which i think you understand already, is that:
1. Data updated by a single instruction can be indexed in any order at any time.
2. Data updated by multiple instructions need to be protected by a sequence number for that specific data, and all instructions that update that particular data must update that data's sequence number.